### ISSUE TITLE
nullptr dereference

### ISSUE THREAD

**Issue by benschlueter** [IAIK/msrevelio#1]
I experience a nullptr dereference when doing the scan. I strictly followed the instructions.

OS: Ubuntu
Kernel: vanilla 5.18 (with Ubuntu config and some tweaks to boot without initramfs)
CPU: Dual Socket 3rd gen scalable Xeon (Gold 6346 CPU @ 3.10GHz) 

```
[ 2446.217858] [MSR_SCAN]: init
[ 2446.218168] [MSR_SCAN]: done
[ 2523.946478] [MSR_SCAN]: inside open
[ 2564.388585] [MSR_SCAN]: inside open
[ 2564.388627] [MSR_SCAN]: Starting to scan...
[ 2564.391228] BUG: kernel NULL pointer dereference, address: 0000000000000000
[ 2564.391241] #PF: supervisor read access in kernel mode
[ 2564.391246] #PF: error_code(0x0000) - not-present page
[ 2564.391251] PGD 0
[ 2564.391257] Oops: 0000 [#1] PREEMPT SMP NOPTI
[ 2564.391265] CPU: 16 PID: 7354 Comm: kworker/u131:1 Tainted: G           OE     5.18.0 #14
[ 2564.391273] Hardware name: Supermicro X12DAi-N6/X12DAi-N6, BIOS 1.1b 09/10/2021
[ 2564.391277] Workqueue: msr_scan_wq scan_msrs [msrs]
[ 2564.391294] RIP: 0010:native_write_msr_safe+0x4/0x30
[ 2564.391312] Code: 5c 5d c3 48 c1 e2 20 49 89 d4 31 d2 49 09 c4 4c 89 e6 e8 0f 88 6c 00 4c 89 e0 41 5c 5d c3 0f 1f 84 00 00 00 00 00 89 f0 89 f9 <0f> 30 31 c0 66 90 c3 55 48 c1 e2 20 89 f6 48 09 d6 89 c2 48 89 e5
[ 2564.391319] RSP: 0018:ff3d936a28e17ca8 EFLAGS: 00010246
[ 2564.391327] RAX: 0000000000000000 RBX: 0000000000000079 RCX: 0000000000000079
[ 2564.391332] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000079
[ 2564.391335] RBP: ff3d936a28e17e30 R08: 000000000000000c R09: 8080808080808080
[ 2564.391340] R10: ff1bf0ee1cfb5034 R11: 000000000000000c R12: 0000000000000079
[ 2564.391343] R13: 0000000000000000 R14: 000000000000c34f R15: 0000000000000024
[ 2564.391348] FS:  0000000000000000(0000) GS:ff1bf10d83a00000(0000) knlGS:0000000000000000
[ 2564.391353] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 2564.391358] CR2: 0000000000000000 CR3: 000000598da10002 CR4: 0000000000771ee0
[ 2564.391362] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[ 2564.391366] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[ 2564.391370] PKRU: 55555554
[ 2564.391373] Call Trace:
[ 2564.391377]  <TASK>
[ 2564.391381]  ? scan_msr_range+0x6a/0x1a0 [msrs]
[ 2564.391401]  scan_msrs+0x72/0xc0 [msrs]
[ 2564.391409]  process_one_work+0x21a/0x3f0
[ 2564.391420]  worker_thread+0x4a/0x3b0
[ 2564.391427]  ? process_one_work+0x3f0/0x3f0
[ 2564.391432]  kthread+0xfd/0x130
[ 2564.391444]  ? kthread_complete_and_exit+0x20/0x20
[ 2564.391454]  ret_from_fork+0x1f/0x30
[ 2564.391470]  </TASK>
[ 2564.391473] Modules linked in: msrs(OE) xt_conntrack xt_MASQUERADE nf_conntrack_netlink nfnetlink xfrm_user xfrm_algo xt_addrtype iptable_filter iptable_nat nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 bpfilter br_netfilter bridge stp llc vmw_vsock_vmci_transport vsock vmw_vmci overlay cuse bnep rndis_host cdc_ether usbnet mii btusb btrtl btbcm btintel btmtk bluetooth ecdh_generic ecc hid_logitech_hidpp input_leds joydev hid_logitech_dj mlx5_ib ib_uverbs ib_core hid_generic usbhid hid intel_rapl_msr intel_rapl_common intel_uncore_frequency intel_uncore_frequency_common i10nm_edac nfit nls_iso8859_1 x86_pkg_temp_thermal intel_powerclamp coretemp mlx5_core kvm_intel kvm snd_hda_codec_realtek snd_hda_codec_generic drm_vram_helper ledtrig_audio snd_hda_codec_hdmi drm_ttm_helper ttm snd_hda_intel snd_intel_dspcfg ipmi_ssif crct10dif_pclmul crc32_pclmul ghash_clmulni_intel aesni_intel crypto_simd cryptd rapl snd_intel_sdw_acpi drm_kms_helper snd_hda_codec mlxfw psample efi_pstore
[ 2564.391619]  snd_hda_core intel_cstate fb_sys_fops tls snd_seq_midi syscopyarea snd_seq_midi_event snd_hwdep sysfillrect snd_pcm sysimgblt pci_hyperv_intf isst_if_mmio snd_rawmidi snd_seq snd_seq_device snd_timer isst_if_mbox_pci isst_if_common igb snd mei_me i2c_algo_bit i2c_i801 ioatdma mei ahci i2c_smbus xhci_pci soundcore libahci xhci_pci_renesas intel_pch_thermal dca wmi acpi_ipmi ipmi_si acpi_power_meter acpi_pad mac_hid sch_fq_codel ipmi_devintf ipmi_msghandler drm msr parport_pc ppdev lp parport ip_tables x_tables autofs4
[ 2564.391715] CR2: 0000000000000000
[ 2564.391721] ---[ end trace 0000000000000000 ]---
[ 2564.458736] RIP: 0010:native_write_msr_safe+0x4/0x30
[ 2564.458752] Code: 5c 5d c3 48 c1 e2 20 49 89 d4 31 d2 49 09 c4 4c 89 e6 e8 0f 88 6c 00 4c 89 e0 41 5c 5d c3 0f 1f 84 00 00 00 00 00 89 f0 89 f9 <0f> 30 31 c0 66 90 c3 55 48 c1 e2 20 89 f6 48 09 d6 89 c2 48 89 e5
[ 2564.458761] RSP: 0018:ff3d936a28e17ca8 EFLAGS: 00010246
[ 2564.458767] RAX: 0000000000000000 RBX: 0000000000000079 RCX: 0000000000000079
[ 2564.458771] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000079
[ 2564.458775] RBP: ff3d936a28e17e30 R08: 000000000000000c R09: 8080808080808080
[ 2564.458779] R10: ff1bf0ee1cfb5034 R11: 000000000000000c R12: 0000000000000079
[ 2564.458782] R13: 0000000000000000 R14: 000000000000c34f R15: 0000000000000024
[ 2564.458787] FS:  0000000000000000(0000) GS:ff1bf10d83a00000(0000) knlGS:0000000000000000
[ 2564.458792] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 2564.458796] CR2: 0000000000000000 CR3: 000000598da10002 CR4: 0000000000771ee0
[ 2564.458801] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[ 2564.458804] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[ 2564.458809] PKRU: 55555554
```

Currently I didn't had time to investigate the issue, but I will take a look now.

**Comment by 0xhilbert** [comment]
Dear Benedict Schlüter!

Thank you for reporting this issue. I took some time to investigate this issue, which was a bit more complex than anticipated.

1) The MSR crashing the system is *MSR_IA32_UCODE_WRITE* (0x79), which is responsible for triggering a ucode update.
2) The MSR expects a valid pointer in the `edx:eax` registers to point to a ucode update blob, which is never the case with our implementation.
3) Why it crashes now: Newer CPU versions handle "invalid" ucode blob pointers differently. Previous generations raised a general protection fault that the `native_write_msr_safe` function can handle. Newer CPUs will raise a page fault which is unhandled in the function.

The Solution:
We skip writing to the *MSR_IA32_UCODE_WRITE MSR* altogether.
This is the simplest solution for now. Future solutions could check for page faults, but this might be more work than adding the documented behavior to the list.

The issue was resolved in  #https://github.com/0xhilbert/msrevelio_scanner/commit/beb4701d2d63b9276eed3a4ea8cd752965b0bb56 and `msr_scanner` submodule was redirected to the patched version for now.

I've tested the patch on an Intel Alder Lake CPU with kernels 5.15 and 5.19.

Thanks - Andreas
